File: python/evaluation/CMDPSimulator.py
#################################################################################
# ConstrainedPlanningToolbox
# Copyright (C) 2019 Algorithmics group, Delft University of Technology
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#################################################################################

from util.ProbabilitySample import ProbabilitySample
from util.ToolboxServer import ToolboxServer

class CMDPSimulator:

    def __init__(self, instance):
        self.instance = instance
        self.mean_total_cost = []
        self.mean_instantaneous_cost = []
        self.violation_prob_estimate_total = []
        self.violation_prob_estimate_instantaneous = []

    def run(self, num_runs):
        ToolboxServer.send_request("startSimulation")

        cmdps = self.instance.cmdps
        num_agents = self.instance.get_num_agents()
        num_decisions = self.instance.num_decisions
        num_domain_resources = self.instance.num_domain_resources

        mean_reward = 0.0
        self.mean_total_cost = [0.0 for k in range(num_domain_resources)]
        self.mean_instantaneous_cost = [[0.0 for t in range(num_decisions)] for k in range(num_domain_resources)]
        num_total_violations = [0 for k in range(num_domain_resources)]
        num_instantaneous_violations = [[0 for t in range(num_decisions)] for k in range(num_domain_resources)]

        for run in range(num_runs):
            run_reward = 0.0
            run_total_cost = [0.0 for k in range(num_domain_resources)]
            run_instantaneous_cost = [[0.0 for t in range(num_decisions)] for k in range(num_domain_resources)]

            # get initial states
            state = [0 for i in range(num_agents)]
            for i in range(num_agents):
                state[i] = cmdps[i].initial_state

            for t in range(num_decisions):
                # ask server for actions and convert to list with ints
                actions = ToolboxServer.send_request("getActionsCMDP_{0}_{1}".format(t, state))
                actions = actions.split()
                for i in range(len(actions)):
                    actions[i] = int(actions[i])

                # execute the actions and sample next states
                for i in range(num_agents):
                    cmdp = cmdps[i]
                    s = state[i]
                    a = actions[i]

                    run_reward += cmdp.get_time_reward(t, s, a)
                    for k in range(num_domain_resources):
                        run_total_cost[k] += cmdp.get_cost(k, s, a)
                        run_instantaneous_cost[k][t] += cmdp.get_cost(k, s, a)

                    destinations = cmdp.get_time_transition_destinations(t, s, a)
                    probabilities = cmdp.get_time_transition_probabilities(t, s, a)

                    ps = ProbabilitySample(destinations, probabilities)
                    state[i] = ps.sample_item()


            mean_reward += (run_reward / num_runs)
            for k in range(num_domain_resources):
                self.mean_total_cost[k] += run_total_cost[k] / num_runs

                if self.instance.use_budget_constraints and run_total_cost[k] > self.instance.get_cost_limit_budget(k):
                    num_total_violations[k] += 1

                for t in range(num_decisions):
                    self.mean_instantaneous_cost[k][t] += run_instantaneous_cost[k][t] / num_runs
                    if (not self.instance.use_budget_constraints) and run_instantaneous_cost[k][t] > self.instance.get_cost_limit_instantaneous(k, t):
                        num_instantaneous_violations[k][t] += 1

        # compute violation probability estimates
        self.violation_prob_estimate_total = [0.0 for k in range(num_domain_resources)]
        self.violation_prob_estimate_instantaneous = [[0.0 for t in range(num_decisions)] for k in range(num_domain_resources)]
        for k in range(num_domain_resources):
            self.violation_prob_estimate_total[k] = num_total_violations[k] / num_runs

            for t in range(num_decisions):
                self.violation_prob_estimate_instantaneous[k][t] = num_instantaneous_violations[k][t] / num_runs

        return mean_reward


    def get_mean_instantaneous_cost(self, k, t):
        return self.mean_instantaneous_cost[k][t]

    def get_mean_total_cost(self, k):
        return self.mean_total_cost[k]

    def get_violation_probability_estimate_total(self, k):
        return self.violation_prob_estimate_total[k]

    def get_violation_probability_estimate_instantaneous(self, k, t):
        return self.violation_prob_estimate_instantaneous[k][t]

File: src/main/java/evaluation/CMDPSimulator.java
/*******************************************************************************
 * ConstrainedPlanningToolbox
 * Copyright (C) 2019 Algorithmics group, Delft University of Technology
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 *******************************************************************************/
package evaluation;

import instances.CMDPInstance;
import instances.ConstraintType;

import java.util.Random;

import solutions.mdp.CMDPSolution;
import util.ProbabilitySample;

import model.MDP;
import model.CMDP;

public class CMDPSimulator {
	private MDP[] mdps;
	private CMDPSolution solution;
	private int nAgents;
	private int numDecisions;
	private int numDomainResources;
	
	private CMDPInstance instance;
	
	private Random rnd;
	
	private boolean simulationDone = false;
	private double meanReward;
	private double[] meanTotalCost;
	private double[][] meanInstantaneousCost;
	private double[] violationProbEstimateTotal;
	private double[][] violationProbEstimateInstantaneous;
	
	public CMDPSimulator(CMDPInstance instance, CMDPSolution solution, Random rnd) {
		this.mdps = instance.getCMDPs();
		this.solution = solution;
		this.nAgents = instance.getCMDPs().length;
		this.rnd = rnd;
		this.numDomainResources = instance.getNumDomainResources();
		this.numDecisions = instance.getNumDecisions();
		this.instance = instance;
	}
	
	/**
	 * Run simulation
	 * @param runs number of runs
	 */
	public void run(int runs) {		
		meanReward = 0.0;
		meanTotalCost = new double[numDomainResources];
		meanInstantaneousCost = new double[numDomainResources][numDecisions];
		
		int[] numTotalViolations = new int[numDomainResources];
		int[][] numInstantaneousViolations = new int[numDomainResources][numDecisions];
		
		ProbabilitySample ps = new ProbabilitySample(rnd);
		
		for(int run=0; run<runs; run++) {
			double runReward = 0.0;
			double[] runTotalCost = new double[numDomainResources];
			double[][] runInstantaneousCost = new double[numDomainResources][numDecisions];
			
			// determine initial states
			int[] state = new int[nAgents];
			for(int i=0; i<nAgents; i++) {
				MDP mdp = mdps[i];
				state[i] = mdp.getInitialState();
			}
			
			for(int t=0; t<numDecisions; t++) {
				// select actions
				int[] actions = solution.getActions(t, state);
				
				// get reward and cost, sample next state
				for(int i=0; i<nAgents; i++) {
					MDP mdp = mdps[i];
					
					if(mdp instanceof CMDP) {
						runReward += mdp.getReward(t, state[i], actions[i]);
						
						for(int k=0; k<numDomainResources; k++) {
							runTotalCost[k] += ((CMDP)mdp).getCost(k, state[i], actions[i]);
							runInstantaneousCost[k][t] += ((CMDP)mdp).getCost(k, state[i], actions[i]);
						}
					}
					else {
						runReward += mdp.getReward(t, state[i], actions[i]);
					}
					
					// sample next state
					ps = new ProbabilitySample(rnd);
					int[] transitionDestinations = mdp.getTransitionDestinations(t, state[i], actions[i]);
					double[] transitionProbabilities = mdp.getTransitionProbabilities(t, state[i], actions[i]);
					
					for(int j=0; j<transitionDestinations.length; j++) {
						int s = transitionDestinations[j];
						double prob = transitionProbabilities[j];
						ps.addItem(s, prob);
					}
					int stateNext = ps.sampleItem();
					assert stateNext >= 0 && stateNext < mdp.getNumStates();
					state[i] = stateNext;
				}
			}
			
			meanReward += runReward / ((double) runs);
			for(int k=0; k<numDomainResources; k++) {
				meanTotalCost[k] += runTotalCost[k] / ((double) runs);
				
				if(instance.getConstraintType() == ConstraintType.BUDGET && runTotalCost[k] > instance.getCostLimit(k)) {
					numTotalViolations[k]++;
				}
				
				for(int t=0; t<numDecisions; t++) {
					meanInstantaneousCost[k][t] += runInstantaneousCost[k][t] / ((double) runs);
					
					if(instance.getConstraintType() == ConstraintType.INSTANTANEOUS && runInstantaneousCost[k][t] > instance.getCostLimit(k, t)) {
						numInstantaneousViolations[k][t]++;
					}
				}
			}
		}
		
		simulationDone = true;
		
		// compute violation probability estimates
		violationProbEstimateTotal = new double[numDomainResources];
		violationProbEstimateInstantaneous = new double[numDomainResources][numDecisions];
		for(int k=0; k<numDomainResources; k++) {
			violationProbEstimateTotal[k] = ((double) numTotalViolations[k]) / ((double) runs);
			
			for(int t=0; t<numDecisions; t++) {
				violationProbEstimateInstantaneous[k][t] = ((double) numInstantaneousViolations[k][t]) / ((double) runs);
			}
		}
	}
	
	/**
	 * Get total mean value, based on simulations starting from the initial state
	 * @return value
	 */
	public double getMeanReward() {
		assert simulationDone;
		return meanReward;
	}
	
	/**
	 * Get the mean cost obtained at time t
	 * @param t time t
	 * @return mean cost
	 */
	public double getMeanInstantaneousCost(int k, int t) {
		assert k>=0 && k<numDomainResources && t>=0 && t<numDecisions;
		assert simulationDone;
		return meanInstantaneousCost[k][t];
	}
	
	/**
	 * Get mean total cost for resource k
	 * @param k resource id
	 * @return mean total cost
	 */
	public double getMeanTotalCost(int k) {
		return meanTotalCost[k];
	}
	
	/**
	 * Get empirical estimate of violation probability of resource k
	 * @param k rsource id
	 * @return violation probability estimate
	 */
	public double getViolationProbabilityEstimateTotal(int k) {
		return violationProbEstimateTotal[k];
	}
	
	/**
	 * Get empirical estimate of violation probability of resource k at time t
	 * @param k resource id
	 * @param t time step
	 * @return violation probability estimate
	 */
	public double getViolationProbabilityEstimateInstantaneous(int k, int t) {
		return violationProbEstimateInstantaneous[k][t];
	}
	
}
